File: core/src/main/java/dev/bee/kanjianki/core/StudyMoreNewCardsPolicy.java
package dev.bee.kanjianki.core;

public final class StudyMoreNewCardsPolicy {
    public static final String NO_NEW_CARDS_AVAILABLE_MESSAGE = "No new cards are available.";
    public static final String WHOLE_NUMBER_ERROR_MESSAGE = "Use a whole number of new cards.";
    public static final String POSITIVE_COUNT_ERROR_MESSAGE = "Use at least 1 new card.";

    private StudyMoreNewCardsPolicy() {
    }

    public static int defaultRequestCount(int availableCount) {
        return Math.max(1, Math.min(5, availableCount));
    }

    public static RequestDecision requestedCount(String rawText) {
        int requested;
        try {
            requested = Integer.parseInt(rawText.trim());
        } catch (NumberFormatException error) {
            return RequestDecision.rejected(WHOLE_NUMBER_ERROR_MESSAGE);
        }
        if (requested <= 0) {
            return RequestDecision.rejected(POSITIVE_COUNT_ERROR_MESSAGE);
        }
        return RequestDecision.accepted(requested);
    }

    public static String partialAvailabilityMessage(int admittedCount) {
        return "Only " + StudyTextCopy.countText(admittedCount, "new card was", "new cards were") + " available.";
    }

    public record RequestDecision(int requestedCount, String message) {
        public boolean accepted() {
            return requestedCount > 0;
        }

        private static RequestDecision accepted(int requestedCount) {
            return new RequestDecision(requestedCount, "");
        }

        private static RequestDecision rejected(String message) {
            return new RequestDecision(-1, message);
        }
    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/StudyMoreNewCardsPolicy.kt
package dev.bee.kanjianki.core

import java.util.Locale
import kotlin.math.max
import kotlin.math.min

object StudyMoreNewCardsPolicy {
    const val NO_NEW_CARDS_AVAILABLE_MESSAGE: String = "No new cards are available."
    const val WHOLE_NUMBER_ERROR_MESSAGE: String = "Use a whole number of new cards."
    const val POSITIVE_COUNT_ERROR_MESSAGE: String = "Use at least 1 new card."

    private const val NO_NEW_CARDS_AVAILABLE_MESSAGE_JAPANESE: String = "新しいカードはありません。"
    private const val WHOLE_NUMBER_ERROR_MESSAGE_JAPANESE: String = "新規カード数は整数で入力してください。"
    private const val POSITIVE_COUNT_ERROR_MESSAGE_JAPANESE: String = "新規カード数は1以上で入力してください。"
    private const val JAPANESE_LANGUAGE = "ja"

    @JvmStatic
    fun noNewCardsAvailableMessage(): String = localizedText(
        NO_NEW_CARDS_AVAILABLE_MESSAGE,
        NO_NEW_CARDS_AVAILABLE_MESSAGE_JAPANESE,
    )

    @JvmStatic
    fun wholeNumberErrorMessage(): String = localizedText(
        WHOLE_NUMBER_ERROR_MESSAGE,
        WHOLE_NUMBER_ERROR_MESSAGE_JAPANESE,
    )

    @JvmStatic
    fun positiveCountErrorMessage(): String = localizedText(
        POSITIVE_COUNT_ERROR_MESSAGE,
        POSITIVE_COUNT_ERROR_MESSAGE_JAPANESE,
    )

    @JvmStatic
    fun defaultRequestCount(availableCount: Int): Int {
        return max(1, min(5, availableCount))
    }

    @JvmStatic
    fun requestedCount(rawText: String): RequestDecision {
        val requested = try {
            rawText.trim().toInt()
        } catch (error: NumberFormatException) {
            return RequestDecision.rejected(wholeNumberErrorMessage())
        }
        if (requested <= 0) {
            return RequestDecision.rejected(positiveCountErrorMessage())
        }
        return RequestDecision.accepted(requested)
    }

    @JvmStatic
    fun partialAvailabilityMessage(admittedCount: Int): String {
        return localizedText(
            "Only " + StudyTextCopy.countText(admittedCount, "new card was", "new cards were") + " available.",
            "新規カードは${admittedCount}件のみ使用できます。",
        )
    }

    private fun localizedText(english: String, japanese: String): String =
        if (isJapaneseLocale()) japanese else english

    private fun isJapaneseLocale(): Boolean = Locale.getDefault().language == JAPANESE_LANGUAGE

    class RequestDecision private constructor(
        private val requestedCount: Int,
        private val message: String,
    ) {
        fun requestedCount(): Int = requestedCount

        fun message(): String = message

        fun accepted(): Boolean = requestedCount > 0

        companion object {
            fun accepted(requestedCount: Int): RequestDecision = RequestDecision(requestedCount, "")

            fun rejected(message: String): RequestDecision = RequestDecision(-1, message)
        }
    }
}
